include/libtorrent/alert_manager.hpp

```diff
@@ -160,16 +223,141 @@ namespace libtorrent {
 		alert_manager(alert_manager const&);
 		alert_manager& operator=(alert_manager const&);
 
-		void maybe_notify(alert* a, mutex::scoped_lock& lock);
+		void init_thread_specific_storage()
+		{
+			thread_state ts;
+			ts.youngest_generation = new boost::atomic<int>(0);
```

Comment:
Because every thread gets it's own allocators so there is no contention. Libtorrent's threads access it through `m_generation()` which is a `boost::thread_specific_ptr` and the user threads access it through a list where there is one `thread_state` for each thead. On a later commit I made the `thread_state` pointer thread specific and libtorrent threads access the generation and allocator directly from it, so there is only one thread specific pointer. That probably makes more sense but I discarded that commit because I was trying to store the alerts on a pool and it didn't work out.

Reply:
If you look at the commit before the one that adds thread specific storage, there I solved the race conditions by using more allocators than threads. That's another possibility, but I had to protect each allocation with a mutex. This should be more efficient.